### DISCUSSION TITLE
v4.0.0

### DISCUSSION THREAD

**Discussion by dewenni** [dewenni/ESP_Buderus_KM271#82] [Announcements]
# v4.0.0

## what's new

This major Update brings a completely new web frontend 🎉
Instead of previous that was based on ESPUI-Library, the new one is build on basic html, css and javascript.
Also the communication between server and client has changed from websocket to SSE (Server-Sent Events) and Fetch-API. 

The new webUI Desktop style:
![image](https://github.com/dewenni/ESP_Buderus_KM271/assets/46074831/7e66e128-1b24-4a29-ba50-200709c8e854)
(Desktop Version)

and in fact that the new WebUI has responsive design, it offers also a mobile layout:

![image](https://github.com/dewenni/ESP_Buderus_KM271/assets/46074831/05a7f7da-ee8c-47e5-9f3f-0ae57804bb85)
(Mobile Version)

### Messaging option - Pushover

There is now the possibility to get messages via Pushover. 
Depending on the parameter "Filter", you can define what kind of messages you want to receive.
In the settings you can find all necessary parameters to setup the client.

Each application, service, or utility that sends notifications through Pushover's API needs to have its own API token which uniquely identifies all of the API requests that it makes.
API tokens are free and can be registered through [Pushover website](https://pushover.net/apps/build).

![image](https://github.com/dewenni/ESP_Buderus_KM271/assets/46074831/bbf7ae6c-c7df-4869-ad01-0db81568aebd)

### WebUI-Logger

There is also a new log function with which you can record various messages depending on the filter and display them via the WebUI. This can be useful for your own debugging and also for the further development of the software.

![image](https://github.com/dewenni/ESP_Buderus_KM271/assets/46074831/14b05e6e-5673-4ce4-8fd3-955c90e1e7a8)

### Telnet

In addition to the WebUI and MQTT, there is also a new Telnet interface to communicate with the ESP.
The interface offers several commands to read out information and send commands.
An overview of the commands can be called up using the "help" command.
To connect, a simple Telnet connection can be started via the corresponding IP address of the ESP.

Example: 
```
> telnet 192.168.178.135
```

![image](https://github.com/dewenni/ESP_Buderus_KM271/assets/46074831/ed1ac8c2-396a-47f1-a3a8-9c4027a20b50)

## changelog

- update to espressif32 6.6.0
- complete new web UI.
- OTA and Config import/export was redesigned
- login authentication was redesigned
- rework messaging function. mqtt publish function was replaced with a generic messaging function for further options.
- add messaging option for "Pushover" #31
- separate mqtt language option independent of webUI language
- add WebUI-Logger #70
- upgrade to ArduinoJSON 7 #71
- bugfix for updating OneWire sensor values if not enabled
- Prevent zero values from being sent at startup for "info = {"burner":0,"pump":100,"ww_temp":57,"boiler_temp":38}"
- improve behavior for incoming mqtt messages without payload (this could lead to problems in the past)
- change DoubleResetDetection time from 10s to 3s
- local copy of ESPAsyncWebServer and AsyncTCP with bugfixes of known issues
- add "Simulation-Mode" for testing purposes. Can be activated in config settings and will provide sample Data of a Logamatic 2107.
- add Telnet Server as an additional operator interface with different command options
- add Version information from github. If you click on the version info on the bottom left, a dialog with the version information of the latest github release is shown.

> [!IMPORTANT]   
> With this version, the dependency of the WebUI language and the MQTT messages has been removed. There is now a separate language setting in the MQTT settings that is used for the MQTT topics!

## Assets

<hr /><em>This discussion was created from the release <a href='https://github.com/dewenni/ESP_Buderus_KM271/releases/tag/v4.0.0'>v4.0.0</a>.</em>

**Comment by dewenni** [comment]
minor bugfix release [release 4.0.1](https://github.com/dewenni/ESP_Buderus_KM271/releases/tag/v4.0.1)

**Comment by dewenni** [comment]
minor bugfix [release v4.0.2](https://github.com/dewenni/ESP_Buderus_KM271/releases/tag/v4.0.2)

**Comment by the78mole** [comment]
Hi @dewenni ,

good news, today, the next batch will start shipping. This time again with 8MB ESP32-flash size. If the user selected you firmware option, it will ship with 4.0.2. Is there anything speaking against it? 

Regards,
Daniel

**Reply by dewenni** [reply]
Hi @the78mole 
For me it's fine!
Hopefully I have fixed all important bugs in my v4 now ;-)

Next step will be the migration to Arduino V3 and support of Ethernet/LAN connection.
